presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -89,400 +111,733 @@
                 node.getOffset(),
                 node.getLimit());
     }
 
     @Override
-    protected Node visitSelect(Select node, Void context)
+    protected Node visitNode(Node node, Void context)
     {
-        if (materializedViewInfo.isDistinct() && !node.isDistinct()) {
-            throw new IllegalStateException("Materialized view has distinct and base query does not");
-        }
-        ImmutableList.Builder<SelectItem> rewrittenSelectItems = ImmutableList.builder();
-
-        for (SelectItem selectItem : node.getSelectItems()) {
-            rewrittenSelectItems.add((SelectItem) process(selectItem, context));
-        }
-
-        return new Select(node.isDistinct(), rewrittenSelectItems.build());
+        return node;
     }
 
     @Override
-    protected Node visitSingleColumn(SingleColumn node, Void context)
+    protected Node visitQuery(Query node, Void context)
     {
-        // For a single table, without sub-queries, the column prefix is unnecessary. Here It is removed so that it can be mapped to the view column properly.
-        // For relations other than single table, it needs to be reserved to differentiate columns from different tables.
-        // One way to do so is to process the prefix within `visitDereferenceExpression()` since the prefix information is saved as `base` in `DereferenceExpression` node.
-        node = removeSingleColumnPrefix(node, removablePrefix);
-        Expression expression = node.getExpression();
-        Optional<Set<Expression>> groupByOfMaterializedView = materializedViewInfo.getGroupBy();
-        // TODO: Replace this logic with rule-based validation framework.
-        if (groupByOfMaterializedView.isPresent() &&
-                validateExpressionForGroupBy(groupByOfMaterializedView.get(), expression) &&
-                (!expressionsInGroupBy.isPresent() || !expressionsInGroupBy.get().contains(expression))) {
-            throw new IllegalStateException("Query a column presents in materialized view group by: " + expression.toString());
-        }
-
-        Expression processedColumn = (Expression) process(expression, context);
-        Optional<Identifier> alias = node.getAlias();
-
-        // If a column name was rewritten, make sure we re-alias to same name as base query
-        if (!alias.isPresent() && processedColumn instanceof Identifier && !processedColumn.equals(node.getExpression())) {
-            alias = Optional.of((Identifier) node.getExpression());
-        }
-        return new SingleColumn(processedColumn, alias);
+        QueryBody newQueryBody = processSameType(node.getQueryBody());
+        Optional<With> newWith = node.getWith().map(this::processSameType);
+        boolean withSame = !node.getWith().isPresent() || (node.getWith().orElseThrow(IllegalStateException::new) == newWith.get());
+        if (withSame && (node.getQueryBody() == newQueryBody)) {
+            return node;
+        }
+        if (node.getLocation().isPresent()) {
```

Comment:
can we be simplified by using the constructor with location, pass empty if not present.

Reply:
Not sure I follow- are you referring to the private constructor referred to in https://github.com/prestodb/presto/pull/17901#discussion_r901841398?

Reply:
Yes, saw it now. We can make the mark optional nodelocation constructor public that should be good enough?

Reply:
My discussion with @pranjalssh regarding this ended up with us deciding this was better for now,

- There are a lot of classes that extend `Node`
- All of them following the convention of having one constructor with location and one without, which both call a private constructor that takes an optional
- We don't like the idea of making just the ones I'm using here have a public constructor. I personally would want to make all `Node`s take an `Optional`, or none.